File: Python/day2-1.py
import re


def isvalidpassword(password: str, char: str, mintimes: int, maxtimes: int):
    times = password.count(char)

    return mintimes <= times <= maxtimes


def processline(line: str):
    template = "(\d*)-(\d*) (\w): (\w*)"

    mintimes, maxtimes, character, password = re.match(template, line).groups()

    mintimes = int(mintimes)
    maxtimes = int(maxtimes)

    return password, character, mintimes, maxtimes


nvalid = 0
with open("Recursos/day2-input.txt") as io:
    for line in io:
        nvalid += isvalidpassword(*processline(line))

print(nvalid)

File: Julia/day2-1.jl
function processline(line::AbstractString)
    re = r"(\d*)-(\d*) (\w): (\w*)"

    mintimes, maxtimes, character, password = match(re, line).captures

    mintimes = parse(Int, mintimes)
    maxtimes = parse(Int, maxtimes)

    return password, character[1], mintimes, maxtimes
end

function isvalidpassword(password::AbstractString, char::AbstractChar,
    mintimes::Integer, maxtimes::Integer)
    times = count(c -> c == char, password)

    return mintimes ≤ times ≤ maxtimes
end


nvalid = 0
open("Recursos/day2-input.txt") do io
    for line in eachline(io)
        password, char, mintimes, maxtimes = processline(line)

        global nvalid += isvalidpassword(password, char, mintimes, maxtimes)
    end
end

println(nvalid)
